suite/volume/volume_setup.go

```diff
@@ -0,0 +1,335 @@
+	if err := retry.Do(func() error {
+		if !vt.checkRoundFinishedOnChain(roundID, newVal) {
+			return errors.New("round is not finished")
+		}
+		return nil
+	}, vt.OnChainCheckAttemptsOpt); err != nil {
+		return errors.Wrap(err, "round is not fully finished on chain")
+	}
+	return nil
+}
+
+func (vt *FluxTest) debugRoundTuple(rounds []*contracts.FluxAggregatorData) {
+	roundValsArr := make([]roundVals, 0)
+	for _, r := range rounds {
+		if r == nil {
+			continue
+		}
+		roundValsArr = append(roundValsArr, roundVals{
+			RoundID: r.LatestRoundData.RoundId.Int64(),
+			Val:     r.LatestRoundData.Answer.Int64(),
+		})
+	}
+	log.Debug().Interface("rounds_on_chain", roundValsArr).Msg("last rounds on chain")
+}
+
+func (vt *FluxTest) checkRoundFinishedOnChain(roundID int, newVal int) bool {
+	log.Debug().Int("round_id", roundID).Msg("checking round completion on chain")
+	var rounds []*contracts.FluxAggregatorData
+	for _, flux := range *vt.FluxInstances {
+		cd, err := flux.GetContractData(context.Background())
+		if err != nil {
+			log.Err(err).Msg("error checking flux last round")
+			return false
+		}
+		rounds = append(rounds, cd)
+	}
+	vt.debugRoundTuple(rounds)
+	for _, r := range rounds {
+		if r.LatestRoundData.RoundId.Int64() != int64(roundID) || r.LatestRoundData.Answer.Int64() != int64(newVal) {
+			return false
+		}
+	}
+	return true
+}
+
+// roundEndTimes counts submissions for round, when all required submission is found, saves last block_timestamp for contract
+func (vt *FluxTest) roundEndTimes(hr *ethlog.HistoryResult, expectedRound int, submitsRequired int) (map[string]int64, error) {
+	endTimesMap := make(map[string]int64)
+	submitsMap := make(map[string]int)
+	for _, block := range hr.History {
+		if len(block["transactions"].([]ethlog.ParsedTx)) == 0 {
+			//log.Debug().Msg("empty block")
+			continue
+		}
+		// for every tx search event with matching round id, happened in particular contract
+		for _, tx := range block["transactions"].([]ethlog.ParsedTx) {
+			for _, event := range tx["events"].([]ethlog.ParsedEvent) {
+				data := event["event_data"].(ethlog.RawParsedEventData)
+				if data["round"] == nil {
+					continue
+				}
+				addr := event["event_address"].(common.Address)
+				// identifying submission by "round" field
+				if int(data["round"].(uint32)) == expectedRound {
+					submitsMap[addr.Hex()] += 1
+				}
+				if submitsMap[addr.Hex()] > submitsRequired {
+					return nil, errors.New(fmt.Sprintf("more that required submits found for contract: %s", addr.Hex()))
+				}
+				if submitsMap[addr.Hex()] == submitsRequired {
+					// milliseconds
+					endTimesMap[addr.Hex()] = int64(block["block_time"].(uint64) * 1000)
+				}
+			}
+		}
+	}
+	log.Debug().Interface("round_end_block_times", endTimesMap).Send()
+	return endTimesMap, nil
+}
+
+// getOnChainLogs get all block data for test interval
+func (vt *FluxTest) getOnChainLogs(from *big.Int, to *big.Int, roundID int) (*ethlog.HistoryResult, error) {
```

Comment:
This is ETH specific, we should have 0 ETH specific logic outside of the contract implementations.